src/main/scala/passes/HoistSlowBinop.scala

```diff
@@ -0,0 +1,122 @@
+package fuselang.passes
+
+import scala.{PartialFunction => PF}
+import fuselang.common._
+import Syntax._
+// import CodeGenHelpers._
+// import CompilerError._
+import Transformer._
+import EnvHelpers._
+
+object HoistSlowBinop extends TypedPartialTransformer {
+
+  case class ExprEnv(map: Map[Expr, CLet])
+      extends ScopeManager[ExprEnv]
+      with Tracker[Expr, CLet, ExprEnv] {
+    def merge(that: ExprEnv) = {
+      ExprEnv(this.map ++ that.map)
+    }
+    def get(key: Expr) = this.map.get(key)
+    def add(key: Expr, value: CLet) = {
+      ExprEnv(this.map + (key -> value))
+    }
+  }
+
+  type Env = ExprEnv
+  val emptyEnv = ExprEnv(Map())
+
+  private val slowBinops = List("*", "/", "%")
+
+  var idx: Map[String, Int] = Map();
+  def genName(base: String): Id = {
+    // update idx
+    idx.get(base) match {
+      case Some(n) => idx = idx + (base -> (n + 1))
+      case None => idx = idx + (base -> 0)
+    }
+    Id(s"$base${idx(base)}_")
+  }
+
+  def construct(
+      cmd: Command,
+      env: Env,
+      acc: Command = CEmpty
+  ): (Command, Env) = {
+    if (env.map.values.isEmpty && acc == CEmpty) {
+      cmd -> emptyEnv
+    } else {
+      CSeq.smart(env.map.values.toSeq :+ acc :+ cmd) -> emptyEnv
+    }
+  }
+
+  def binopRecur(expr: Expr, env: Env): (Expr, Env) = {
+    expr match {
+      // only recur when children are binops
+      case EBinop(_, _, _) => rewriteE(expr)(env)
+      case _ => (expr, env)
+    }
+  }
+
+  def myRewriteE: PF[(Expr, Env), (Expr, Env)] = {
+    case (e @ EBinop(op, left, right), env) if slowBinops.contains(op.op) => {
+      env.get(e) match {
+        case Some(let) => (EVar(let.id), env)
+        case None => {
+          val (leftRead, leftEnv) = binopRecur(left, env)
+          val (rightRead, rightEnv) = binopRecur(right, leftEnv)
+          val let = CLet(
+            genName("bin_read"),
+            None,
+            Some(EBinop(op, leftRead, rightRead))
+          )
+          EVar(let.id) -> rightEnv.add(EBinop(op, leftRead, rightRead), let)
+        }
+      }
+    }
+  }
+
+  def myRewriteC: PF[(Command, Env), (Command, Env)] = {
+    case (CLet(id, typ, Some(e)), _) => {
+      val (expr, env) = rewriteE(e)(emptyEnv)
+      construct(CLet(id, typ, Some(expr)), env)
+    }
+    case (CIf(cond, cons, alt), _) => {
+      val (expr, env) = rewriteE(cond)(emptyEnv)
+      construct(
+        CIf(expr, rewriteC(cons)(emptyEnv)._1, rewriteC(alt)(emptyEnv)._1),
+        env
+      )
+    }
+    case (CWhile(cond, pipeline, body), _) => {
+      val (expr, env) = rewriteE(cond)(emptyEnv)
+      construct(
+        CWhile(expr, pipeline, rewriteC(body)(env)._1),
+        env
+      )
+    }
+    case (CUpdate(lhs, rhs), _) => {
+      val (rewrLhs, env) = rewriteE(lhs)(emptyEnv)
+      val (rewrRhs, nEnv) = rewriteE(rhs)(env)
+      construct(CUpdate(rewrLhs, rewrRhs), nEnv)
+    }
+    // XXX(sam): I probably should deal with `*=`
+    case (CReduce(rop, lhs, rhs), _) => {
```

Comment:
If your code doesn't deal with it, `throw` an error. I'd much rather have errors than silently generated bad code.